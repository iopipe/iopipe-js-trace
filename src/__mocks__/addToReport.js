export const invocations = [];

export function addToReport(pluginInstance, timelineArg) {
  const { timeline, invocationInstance } = pluginInstance;
  const entries = (timelineArg || timeline).getEntries();
  const { report } = invocationInstance.report;
  report.performanceEntries = (report.performanceEntries || []).concat(entries);
  invocations.push(invocationInstance);
}
