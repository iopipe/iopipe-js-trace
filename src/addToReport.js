import { unflatten } from 'flat';

function headersObjToArray(headerObj) {
  return Object.keys(headerObj || {}).map(key => {
    return { key, string: headerObj[key] };
  });
}

export function addToReport(pluginInstance, timelineArg) {
  const { timeline, invocationInstance } = pluginInstance;
  const entries = (timelineArg || timeline).getEntries();
  const { report } = invocationInstance.report;
  report.performanceEntries = (report.performanceEntries || []).concat(entries);
}

export function addHttpTracesToReport(plugin) {
  const { autoHttpData: { timeline = {} } } = plugin;
  const { report: { report = {} } = {} } = plugin.invocationInstance;
  Object.keys(plugin.autoHttpData.data).forEach(id => {
    const obj = unflatten(plugin.autoHttpData.data[id] || {});
    if (obj.request) {
      obj.request.headers = headersObjToArray(obj.request.headers);
    }
    if (obj.response) {
      obj.response.headers = headersObjToArray(obj.response.headers);
    }
    // use start mark for startTime in case the http call did not finish / no callback
    // and we do not have a measurement
    const [startMark = {}] = timeline.getEntriesByName(`start:${id}`) || [];
    const [measureMark = {}] = timeline.getEntriesByName(`measure:${id}`) || [];
    obj.timestamp = startMark.timestamp || 0;
    obj.startTime = startMark.startTime || 0;
    obj.duration = measureMark.duration || 0;
    report.httpTraceEntries.push(obj);
  });
}
