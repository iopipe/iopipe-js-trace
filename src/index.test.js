import _ from 'lodash';
import delay from 'delay';
import iopipe from '@iopipe/core';
import mockContext from 'aws-lambda-mock-context';
import Perf from 'performance-node';
import pkg from '../package';
import { invocations } from './addToReport';

const tracePlugin = require('.');

jest.mock('./addToReport');

test('Can instantiate the plugin with no options', () => {
  const plugin = tracePlugin();
  const inst = plugin({});
  expect(_.isFunction(inst.hooks['post:setup'])).toBe(true);
  expect(_.isFunction(inst.postSetup)).toBe(true);
  expect(_.isFunction(inst.hooks['pre:report'])).toBe(true);
  expect(_.isFunction(inst.preReport)).toBe(true);
  expect(_.isPlainObject(inst.config)).toBe(true);
  expect(inst.timeline instanceof Perf).toBe(true);
  expect(inst.config.autoMeasure).toBe(true);
  expect(inst.meta.name).toBe('@iopipe/trace');
  expect(inst.meta.version).toBe(pkg.version);
  expect(inst.meta.homepage).toBe(pkg.homepage);
});

test('Works with iopipe', async () => {
  try {
    const iopipeInstance = iopipe({ token: 'test', plugins: [tracePlugin()] });
    let testStartDate, testEndDate;
    const wrappedFn = iopipeInstance(async (event, context) => {
      const { mark, measure } = context.iopipe;
      testStartDate = Date.now() - 1;
      mark.start('test');
      await delay(10);
      mark.start('extra');
      mark.end('test');
      await delay(10);
      mark.end('extra');
      measure('custom-measure', 'test', 'extra');
      testEndDate = Date.now() + 1;
      context.succeed('wow');
    });
    const context = mockContext({ functionName: 'test-1a' });
    wrappedFn({}, context);

    const val = await context.Promise;
    expect(val).toBe('wow');

    // run 2 invocations, to more accurately determine trace durations
    const context2 = mockContext({ functionName: 'test-1b' });
    wrappedFn({}, context2);

    await context2.Promise;
    const report = _.chain(invocations)
      .find(obj => obj.context.functionName === 'test-1b')
      .get('report')
      .value();

    const { performanceEntries, labels } = report.report;
    expect(labels.includes('@iopipe/plugin-trace')).toBe(true);
    expect(performanceEntries).toHaveLength(7);
    const [startMark, customMeasure, measure, endMark] = performanceEntries;
    expect(_.inRange(measure.duration, 5, 20)).toBe(true);
    expect(_.isNumber(startMark.timestamp)).toBe(true);
    expect(startMark.timestamp > 15e11).toBe(true);
    expect(_.inRange(startMark.timestamp, testStartDate, testEndDate)).toBe(
      true
    );
    expect(_.inRange(endMark.timestamp, testStartDate, testEndDate)).toBe(true);
    expect(_.inRange(customMeasure.duration, 5, 29)).toBe(true);
  } catch (err) {
    throw err;
  }
});

test('Can disable autoMeasure', async () => {
  try {
    const iopipeInstance = iopipe({
      token: 'test',
      plugins: [tracePlugin({ autoMeasure: false })]
    });
    const wrappedFn = iopipeInstance((event, context) => {
      const { mark } = context.iopipe;
      mark.start('test');
      mark.end('test');
      context.succeed('wow');
    });
    const context = mockContext({ functionName: 'test-2' });
    wrappedFn({}, context);
    await context.Promise;
    const performanceEntries = _.chain(invocations)
      .find(obj => obj.context.functionName === 'test-2')
      .get('report.report.performanceEntries')
      .value();
    expect(performanceEntries).toHaveLength(2);
    const measure = performanceEntries.find(
      item => item.entryType === 'measure'
    );
    expect(measure).toBeUndefined();
  } catch (err) {
    throw err;
  }
});
