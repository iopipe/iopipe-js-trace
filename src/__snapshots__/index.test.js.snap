// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autoHttp works with consecutive invocations 1`] = `
Array [
  Object {
    "request": Object {
      "hash": null,
      "headers": Array [
        Object {
          "key": "user-agent",
          "string": "iopipe-test",
        },
      ],
      "hostname": "www.iopipe.com",
      "method": "GET",
      "path": "/?run=1",
      "pathname": "/",
      "port": null,
      "protocol": "https:",
      "query": "?run=1",
      "url": "https://www.iopipe.com/?run=1",
    },
    "response": Object {
      "headers": Array [
        Object {
          "key": "vary",
          "string": "Accept-Encoding",
        },
        Object {
          "key": "server",
          "string": "Netlify",
        },
        Object {
          "key": "content-type",
          "string": "text/html; charset=UTF-8",
        },
      ],
      "statusCode": 200,
      "statusMessage": "OK",
    },
  },
  Object {
    "request": Object {
      "hash": null,
      "headers": Array [
        Object {
          "key": "user-agent",
          "string": "iopipe-test",
        },
      ],
      "hostname": "www.iopipe.com",
      "method": "GET",
      "path": "/?run=2",
      "pathname": "/",
      "port": null,
      "protocol": "https:",
      "query": "?run=2",
      "url": "https://www.iopipe.com/?run=2",
    },
    "response": Object {
      "headers": Array [
        Object {
          "key": "vary",
          "string": "Accept-Encoding",
        },
        Object {
          "key": "server",
          "string": "Netlify",
        },
        Object {
          "key": "content-type",
          "string": "text/html; charset=UTF-8",
        },
      ],
      "statusCode": 200,
      "statusMessage": "OK",
    },
  },
]
`;

exports[`autoHttp works with got(url) and options 1`] = `
Array [
  Object {
    "request": Object {
      "headers": Array [
        Object {
          "key": "user-agent",
          "string": "iopipe-test",
        },
      ],
      "hostname": "www.iopipe.com",
      "method": "GET",
      "path": "/?test=foo",
      "pathname": "/",
      "port": "443",
      "protocol": "https:",
      "query": "?test=foo",
      "url": "https://www.iopipe.com:443/?test=foo#wowhash",
    },
    "response": Object {
      "headers": Array [
        Object {
          "key": "vary",
          "string": "Accept-Encoding",
        },
        Object {
          "key": "server",
          "string": "Netlify",
        },
        Object {
          "key": "content-type",
          "string": "text/html; charset=UTF-8",
        },
      ],
      "statusCode": 200,
      "statusMessage": "OK",
    },
  },
]
`;

exports[`autoHttp works with got(url) plain 1`] = `
Object {
  "request": Object {
    "hash": "#wowhash",
    "headers": Array [
      Object {
        "key": "user-agent",
        "string": "iopipe-test",
      },
    ],
    "hostname": "www.iopipe.com",
    "method": "GET",
    "path": "/?test=foo",
    "pathname": "/",
    "port": "443",
    "protocol": "https:",
    "query": "?test=foo",
    "url": "https://www.iopipe.com:443/?test=foo#wowhash",
  },
  "response": Object {
    "headers": Array [
      Object {
        "key": "vary",
        "string": "Accept-Encoding",
      },
      Object {
        "key": "server",
        "string": "Netlify",
      },
      Object {
        "key": "content-type",
        "string": "text/html; charset=UTF-8",
      },
    ],
    "statusCode": 200,
    "statusMessage": "OK",
  },
}
`;
